### PR TITLE
Fishes Hatchet Changes

### PR DESCRIPTION
**Pull request by flashfish0x**
### Added reentry checks

Added reentry checks to critical functions

### Added force revoke strategy

Force revoke allows a strategy to be revoked even if it is non-compliant. It sets the remaining assets in the strategy to 0. A new parameter in the revokeStrategy event shows any loss taken to the vault by revoking a strategy. 

TODO. Is it worth DCAing the loss? by burning shares (if there are any to burn)

### Removed Tend and Migrate

Tend and migrate aren't erc4626 compliant and they both require the vault to have special permissions on the strategy. They are also no needed as the same change can be achieved by withdrawing and redepositing elsewhere. If migrate is needed it should be handled at strategy level.

Some more discussion is required here. As migrate is very useful when it costs to withdraw. 

### Change process_report role to KEEPER

Previously process_report was permissioned by ACCOUNTING_MANAGER. A too senior role.

### Queue manager 

Added a new periphery contract called queue_manager that sets the queue on a redeem. This allows v3 to act the same as v2 on withdrawals if we require. It also makes it more erc4626 compliance as with a queue manager . 

TODO: change maxWithdraw to query queue manager

### CODE REVIEW

File: contracts/VaultV3.vy [hunk 1]
```diff
@@ -80,6 +82,9 @@ event DebtUpdated:
 event UpdateAccountant:
     accountant: address
 
+event UpdateQueueManager:
+    queue_manager: address
```
Comment (Schlagonia) [inline contracts/VaultV3.vy]: should it be indexed?

File: contracts/VaultV3.vy [hunk 2]
```diff
@@ -487,10 +493,16 @@ def _assess_share_of_unrealised_losses(strategy: address, assets_needed: uint256
...
 @internal
-def _redeem(sender: address, receiver: address, owner: address, shares_to_burn: uint256, strategies: DynArray[address, 10] = []) -> uint256:
+def _redeem(sender: address, receiver: address, owner: address, shares_to_burn: uint256, _strategies: DynArray[address, 10] = []) -> uint256:
     if sender != owner:
         self._spend_allowance(owner, sender, shares_to_burn)
 
+    strategies: DynArray[address, 10] = _strategies
+
+    queue_manager: address = self.queue_manager
+    if queue_manager != empty(address):
```
Comment (Schlagonia) [inline contracts/VaultV3.vy:516]: This gets rid of the potential for anyone to pass a custom withdrawal queue in if there is a queue manager. Should it be if queue_manager != empty(address) and _strategies == []: 

That way someone can still customize the queue and since all strategies are permisioless anyway it shouldnt add extra risk

File: contracts/VaultV3.vy [hunk 3]
```diff
@@ -15,12 +15,13 @@ interface IStrategy:
...
     def convertToAssets(shares: uint256) -> (uint256): view
     def convertToShares(assets: uint256) -> (uint256): view
-    def migrate(strategy: address): nonpayable
-    def tend(): nonpayable
 
 interface IAccountant:
     def report(strategy: address, gain: uint256, loss: uint256) -> (uint256, uint256): nonpayable
 
+interface IQueueManager:
+    def withdraw_queue() -> (DynArray[address, 10]): nonpayable
```
Comment (Schlagonia) [inline contracts/VaultV3.vy]: with a seperate QueueManager contract we would probably want to use the same one for the vault, a potential router, or any direct front end integration etc. So would want it to take the address of vault as a parameter so anyone can view the queue

File: contracts/VaultV3.vy (hunk 1 above)
Comment (flashfish0x) [reply contracts/VaultV3.vy]: yes. All the events need to be reviewed and indexes added

File: contracts/VaultV3.vy (hunk 2 above)
Comment (flashfish0x) [reply contracts/VaultV3.vy:516]: It's a good question worth discussing. How much do we like people being able to choose custom queues. i think you are right and that the move to permissionless strategies means that security from queue ordering matters a lot less

File: contracts/VaultV3.vy [hunk 4]
```diff
@@ -598,9 +610,14 @@ def _add_strategy(new_strategy: address):
...
-def _revoke_strategy(old_strategy: address):
+def _revoke_strategy(old_strategy: address, force: bool):
    assert self.strategies[old_strategy].activation != 0, "strategy not active"
-   assert self.strategies[old_strategy].current_debt == 0, "strategy has debt"
+   loss: uint256 = 0
+
+   if self.strategies[old_strategy].current_debt != 0:
+    assert force, "strategy has debt"
+    loss = self.strategies[old_strategy].current_debt
+
```
Comment (Schlagonia) [inline contracts/VaultV3.vy:635]: need to decrease self.total_debt by loss amount here right

File: contracts/VaultV3.vy (hunk 2 above)
Comment (flashfish0x) [reply contracts/VaultV3.vy:516]: Also should there be a call to queue manager when a new strategy is added or revoked?

File: contracts/VaultV3.vy (hunk 4 above)
Comment (flashfish0x) [reply contracts/VaultV3.vy:635]: https://github.com/flashfish0x/yearn-vaults-v3/pull/1/commits/13f0152deac86c0041ddce48f97557f0709f58e6

File: contracts/VaultV3.vy (hunk 3 above)
Comment (flashfish0x) [reply contracts/VaultV3.vy]: https://github.com/flashfish0x/yearn-vaults-v3/pull/1/commits/13f0152deac86c0041ddce48f97557f0709f58e6

File: contracts/VaultV3.vy (hunk 1 above)
Comment (flashfish0x) [reply contracts/VaultV3.vy]: https://github.com/flashfish0x/yearn-vaults-v3/pull/1/commits/13f0152deac86c0041ddce48f97557f0709f58e6

File: contracts/VaultV3.vy (hunk 2 above)
Comment (Schlagonia) [reply contracts/VaultV3.vy:516]: I dont think it should call queue manager, the queue can only hold 10 strategies but the vault can hold as many as we want. The queue should be set by gov i would think. To line it up with the expected apr as well as how liquid a strategy is etc. for the easiest cheapest withdraws

Comment (flashfish0x) [reply contracts/VaultV3.vy:516]: Sure, I suppose you could implement it in the adder. Addstrategy + and addToQueue on the queuemanager

Comment (flashfish0x) [reply contracts/VaultV3.vy:516]: > This gets rid of the potential for anyone to pass a custom withdrawal queue in if there is a queue manager. Should it be if queue_manager != empty(address) and _strategies == []:
> 
> That way someone can still customize the queue and since all strategies are permisioless anyway it shouldnt add extra risk

What do you think to passing the strategy list in as a parameter to queue manager as well? That way the queue manager lets the user decide  if they can override the queue for max optionality

Comment (Schlagonia) [reply contracts/VaultV3.vy:516]: > > This gets rid of the potential for anyone to pass a custom withdrawal queue in if there is a queue manager. Should it be if queue_manager != empty(address) and _strategies == []:
> > That way someone can still customize the queue and since all strategies are permisioless anyway it shouldnt add extra risk
> 
> What do you think to passing the strategy list in as a parameter to queue manager as well? That way the queue manager lets the user decide if they can override the queue for max optionality

Wouldnt that add alot of gas to be passing a 10 item array back and forth twice.

Vyper will pass the full array even if only one strategy is included plus the cost to potentially iterate over it and the stored array in the queue manager could add quite a bit of extra.

File: contracts/VaultV3.vy [hunk 5]
```diff
@@ -907,6 +893,12 @@ def set_accountant(new_accountant: address):
     self.accountant = new_accountant
```
Comment (Schlagonia) [inline contracts/VaultV3.vy:907]: I believe it was discussed to not allow the accounting manager to set the accountant since it is a high risk contract right? Should it be changed to the role manager

Comment (flashfish0x) [reply contracts/VaultV3.vy:907]: Is there any downside to adding a bunch more roles? What about having one role per setter. eg

Set_Deposit_Limit_Role
Set_Accountant_Role
Revoke_Strategy_Role

File: contracts/VaultV3.vy (hunk 2 above)
Comment (flashfish0x) [reply contracts/VaultV3.vy:516]: Yeh it would good point. What about a query to a shouldOverride view function in the queue manager? 

I am thinking that can be cheap because the queue manager doesnt need to use a storage slot to return the answer. it can just be a constant which would mean just a basic call

Comment (Schlagonia) [reply contracts/VaultV3.vy:516]: So your thinking if the queue was empty it will just pull from the manager but if not it will check with the manager whether or not to allow the user to override? And if false then pull from the manager anyway

-> " the queue manager doesnt need to use a storage slot to return the answer."

Wouldnt this mean each vault would need its own queue_manager deployed specifically for it? 

I would imagine we would want to use the same queue manager for most of the vaults. Not a big issue to pull one bool from storage either way though

Comment (flashfish0x) [reply contracts/VaultV3.vy:516]: Imagine a queue manager which always override. then the function would just return true regardless.

Comment (flashfish0x) [reply contracts/VaultV3.vy:516]: https://github.com/flashfish0x/yearn-vaults-v3/pull/1/commits/59b85ca95d81f44e0c81e2cab827a2755a3895c2
